### ISSUE TITLE
Python 3: TypeError thrown when comparing different types, especially comparing None versus something else

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#9107]
Hi,

The following not only affects Core, but also add-ons as well:

### Background:
In Python 2, it was possible to perform comparisons between None and another type with no errors. However, Python 3 does not allow this, throwing type error. This is most noticeable if properties of objects that starts out as None (such as obj.value) is compared against an integer or other type.

### Steps to reproduce:
Run the following code fragment in Python 2 and 3 interpreters:

None > 0

### Actual behavior:
In Python 3, TypeError is raised:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'NoneType' and 'int'

Whereas Python 2 returns False.

### Expected behavior:
See remark.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
Alpha.16477 (Python 2 and Python 3 source code)

#### Windows version:
Windows 10 Version 1809

#### Name and version of other software in use when reproducing the issue:
Python 2.7.15, 3.7.2

#### Other information about your system:
Python 3 transition research being tested

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA?
Yes - 2018.3, 2018.4

### Remark:
The most appropriate solution is disallowing comparisons between None and other types, or for that matter, comparisons between different types. One way is instance checking like so:

if isinstance(something, int) and something {operator} other: code

Another strategy is checking if something is None and return if further comparisons do not make sense like so:

if something is None: return default

Solution 1 will enforce type checking while writing and is easier to read, whereas solution 2 may yield slightly faster code if done correctly. I propose looking at type comparisons and choosing the appropriate solution on a case-by-case basis (with comments justifying why).

Thanks.

**Comment by leonardder** [comment]
Are these cases caught by scripts like 2to3 and modernize?

**Comment by josephsl** [comment]
Hi, at least it isn’t caught by 2to3, as they are runtime errors. Thanks.

**Comment by leonardder** [comment]
I've found at least one instance of this while navigating to the Edge-like container in the Windows start menu.

```
ERROR - scriptHandler.executeScript (18:16:09.125):
error executing script: <bound method EditableText.script_caret_moveByLine of <NVDAObjects.Dynamic_StartMenuSearchFieldSearchFieldEditableTextWithoutAutoSelectDetectionUIA object at 0x04488E30>> with gesture 'down arrow'
Traceback (most recent call last):
  File "scriptHandler.py", line 189, in executeScript
    script(gesture)
  File "editableText.py", line 187, in script_caret_moveByLine
    self._caretMovementScriptHelper(gesture, textInfos.UNIT_LINE)
  File "editableText.py", line 146, in _caretMovementScriptHelper
    self._caretScriptPostMovedHelper(unit,gesture,newInfo)
  File "editableText.py", line 130, in _caretScriptPostMovedHelper
    speech.speakTextInfo(info, unit=speakUnit, reason=controlTypes.REASON_CARET)
  File "speech\__init__.py", line 768, in speakTextInfo
    textWithFields=info.getTextWithFields(formatConfig)
  File "NVDAObjects\UIA\__init__.py", line 647, in getTextWithFields
    fields=list(self._getTextWithFieldsForUIARange(self.obj.UIAElement,self._rangeObj,formatConfig))
  File "NVDAObjects\UIA\__init__.py", line 637, in _getTextWithFieldsForUIARange
    for field in self._getTextWithFields_text(textRange,formatConfig):
  File "NVDAObjects\UIA\__init__.py", line 461, in _getTextWithFields_text
    field=self._getFormatFieldAtRange(tempRange,formatConfig,ignoreMixedValues=len(furtherUIAFormatUnits)==0)
  File "NVDAObjects\UIA\__init__.py", line 238, in _getFormatFieldAtRange
    if UIAHandler.StyleId_Heading1<=styleIDValue<=UIAHandler.StyleId_Heading9:
TypeError: '<=' not supported between instances of 'int' and 'POINTER(IUnknown)'

```

```

**Comment by feerrenrut** [comment]
I think this issue can be closed with #9794. Specific instances of these errors are unknown, please raise an issue if you encounter any further instances. Other wise, perhaps type hinting can be used to track these down.

**Comment by josephsl** [comment]
Hi, yep, let’s close this – anyone encountering more issues like this should file new issues. Thanks.
